### ISSUE TITLE
"Blocking waiting for file lock on the git checkout"

### ISSUE THREAD

**Issue by SimonSapin** [rust-lang/cargo#2627]
TL;DR: Cargo seems to block with this message more than necessary. Are the locks ever released before process exit?

---

Servo builds can take a long time, so I have a multiple clones to work on something else while something is building: `~/projects/servo`, `~/projects/servo1`, `~/projects/servo2`. They all share the same `$CARGO_HOME`, but have separate `$CARGO_TARGET_DIR`.

Lately (maybe this behavior has been around for a long time and I only noticed it recently) when running `cargo build` in more than one clone at a time, I see this message "Blocking waiting for file lock on the git checkout". Cargo seems to block until the other build is completely finished. This is unfortunate since building Servo is often a single-core affair since the `script` crate is very large and close to the root of the dependency tree.

The code taking these lock seems to be `cargo::sources::git::GitSource::update`. But they are apparently held a lot longer than the `Updating` steps, possibly until `cargo build` terminates completely.

**Comment by alexcrichton** [comment]
The reason for this is [documented here](https://github.com/rust-lang/cargo/blob/27161ec54540acec41280b5c715058e3c238dfaf/src/cargo/sources/git/source.rs#L199-L201), which is that Cargo locks the checkout for the duration that the `Source` is alive, because the same checkout could be checked out at multiple revisions (e.g. the checkout is just the `master` branch, and multiple repos could check out multiple commits on the master branch).

We could try to employ another strategy, however, along the lines of:

1. Drop sources when they're no longer needed. This'd drop the lock, but may be difficult to figure out *when* it can be done.
2. Optimize the case that you're trying to get the same revision, so allow both to proceed somehow. This may involve a downgrade from a writable to readable lock, which I believe isn't available as a primitive.

For now I'd probably recommend having separate cargo home directories unfortunately.

**Comment by SimonSapin** [comment]
Yeah, I’ve switched to separate `CARGO_HOME`. I was also seeing a lot of things being rebuilt when I didn’t expect it, but that also explains it: I suppose the build of a git dependency is considered out of date if a separate build checked out a different revision in the mean time.

Would it be possible to have separate checkouts for separate revisions of the same git dependency? That’s what happens for crates.io dependencies.

**Comment by alexcrichton** [comment]
Yeah we do have separate checkouts for separate *channels* for a git dependency. For example the `master` branch and the `foo` branch would have different checkouts. The conflicts happen when multiple projects use the same branch (e.g. everyone's using `master`) but different points along that branch.

We could in theory check it out in a different location each time but that may just end up making the cache of git repos a little *too* large :(

**Comment by SimonSapin** [comment]
What makes it not too large for crates.io dependencies?

**Comment by alexcrichton** [comment]
crates.io never changes, just a different caching model

**Comment by SimonSapin** [comment]
Version 1.0.0 of `url` on crates.io never changes, but neither does git commit d5a84c928603ef84fff8a3e0f33b4e119f1f01a2.

`$CARGO_HOME/registry/src/github.com-88ac128001ac3a9a` contains a directory for *each version* of each package/crate ever depended on by something I built, and they never get cleaned up. It gets large, but still manageable. Why would it be different for git dependencies? You don’t need to check out every commit in a repository, only those being built. I don’t believe that git dependencies get `cargo update`d much more frequently than crates.io dependencies in practice.

**Comment by SimonSapin** [comment]
Alternatively, could Cargo stop relying on filesystem timestamps (does it?) for deciding whether to rebuild things whose source is in `$CARGO_HOME`? Relying on version numbers / revision hashes instead.

**Comment by alexcrichton** [comment]
> Why would it be different for git dependencies?

The assumption is that git dependencies *are* updated more frequently than crates.io dependencies, so caching all checkouts would be a waste of space. This is what bundler does, and we'd basically have to measure a difference one way or another to prove them wrong and change the behavior.

> Alternatively, could Cargo stop relying on filesystem timestamps (does it?) for deciding whether to rebuild things whose source is in $CARGO_HOME?

Cargo should already do this, if it's not then that's probably a bug or I'm forgetting something.

**Comment by larsbergstrom** [comment]
We seem to be hitting this intermittently  on one of our CI builders. Trying to remove `~/.cargo` to see if that fixes it. Log:
```
 (view as text)

./mach build --release
 in dir /Users/servo/buildbot/slave/mac-rel-css/build (timeout 1200 secs)
 watching logfiles {}
 argv: ['./mach', 'build', '--release']
 environment:
  CARGO_HOME=/Users/servo/.cargo
  CCACHE=/usr/local/bin/ccache
  HOME=/Users/servo
  LOGNAME=servo
  PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
  PWD=/Users/servo/buildbot/slave/mac-rel-css/build
  RUST_BACKTRACE=1
  SERVO_CACHE_DIR=/Users/servo/.servo
  SHELL=/bin/bash
  TMPDIR=/var/folders/vy/3t5lvfrn2flfkxfn1fmp9zb80000gp/T/
  USER=servo
  XPC_FLAGS=0x0
  XPC_SERVICE_NAME=buildslave
  __CF_USER_TEXT_ENCODING=0x1F6:0x0:0x0
 using PTY: False
    Blocking waiting for file lock on the git checkout

command timed out: 1200 seconds without output running ['./mach', 'build', '--release'], attempting to kill
process killed by signal 9
program finished with exit code -1
elapsedTime=1207.926774

```

**Comment by alexcrichton** [comment]
@larsbergstrom Cargo in theory isn't running concurrently, right? That is, even if you have concurrent builds, all Cargo home dir and output is scoped to the build directory?

**Comment by larsbergstrom** [comment]
@alexcrichton It definitely should not be! My suspicion is that the machine got rebooted or something got aborted in the middle of a build and left the .cargo directory in a bad state. I've purged the directory and will check the logs more closely if this recurs!

**Comment by alexcrichton** [comment]
In theory that's no bad state because there's no on-disk state for these locks (they're cleaned up when a process dies). That's the theory at least...

I'll see if I can reproduce locally

**Comment by SimonSapin** [comment]
> there's no on-disk state for these locks (they're cleaned up when a process dies)

What happens if a process is terminated (e.g. by a power cut) without an opportunity to clean up?

**Comment by alexcrichton** [comment]
Sorry, to clarify I'm not saying "in theory" because it's some weird thing that Cargo does, I'm saying that because Cargo uses the [fs2 crate](https://github.com/danburkert/fs2-rs)'s [locking implementation](https://github.com/danburkert/fs2-rs/blob/master/src/unix.rs#L24-L51) which translates to [`flock`](http://linux.die.net/man/2/flock). This implementation of file locking associates locks with file descriptors, so unless there's a bug in the OS then power being cut doesn't mean anything because these aren't preserved on disk, they're just in-memory locks.

**Comment by SimonSapin** [comment]
I see, thanks. I didn’t know `flock` was a thing and assumed locking was based on the existence of a file.

**Comment by alexcrichton** [comment]
@larsbergstrom do you have some raw buildbot logs I could poke around as well?

**Comment by larsbergstrom** [comment]
This is the only one that I could easily find:
http://build.servo.org/builders/mac-rel-css/builds/1210

Note that I suspect this is an issue with the upgrade that happened on this machine. We had a TON of other python-related failures. I unfortunately already requested to have this machine reinstalled from scratch with 10.11. However, I may be able to reproduce the environment if we try to upgrade another one of our identical builders :-)

**Comment by alexcrichton** [comment]
Ok, I was unfortunately unable to reproduce as well, but if you can reproduce I'd be interesting in confirming that Cargo is indeed stuck acquiring a lock (as opposed to just doing something else after it acquired a lock) or also how to reproduce at all I guess :P

**Comment by larsbergstrom** [comment]
@alexcrichton Sorry I didn't keep the machine around! I'll try to make sure that the next machine we take offline for an upgrade I break in a similar fashion first :-)

CC: @edunham

**Comment by Manishearth** [comment]
> This may involve a downgrade from a writable to readable lock, which I believe isn't available as a primitive.

Why not use LOCK_SH and LOCK_EX as a read-write lock?

It can be made finer by only locking when you're actually using the source, but this needs changes to cargo so that it can re-fetch things and might not work well.

**Comment by alexcrichton** [comment]
At some point you may need to update the checkout, requiring an exclusive lock. Without a downgrade primitive, you can't atomically go to a shared lock. That means you'd have to spin in a loop first acquiring a shared lock, checking if it's at the right rev, and if not dropping the lock for an exclusive version to do the update. You'd then just spin in that loop until your shared lock shows that it's at the right revision.

**Comment by Manishearth** [comment]
How's that worse than what we have now? We unconditionally spin if cargo is doing anything.

**Comment by alexcrichton** [comment]
We don't spin now and an implementation like that isn't guaranteed to ever make progerss

**Comment by Manishearth** [comment]
Why not? Try every 10 seconds to acquire a w lock, and if you can't then sleep for a bit and try again. It will eventually succeed when the r locks are released by other cargos.

I get that there's a downgrade issue, but you can build a downgrade primitive out of what we already have using two locks -- one read write lock as usual and a second write-only lock that is acquired only when you intend to acquire or downgrade the other lock in any form.

**Comment by alexcrichton** [comment]
Right we can technically do whatever we want, but if it's unduly complicated then it's pretty likely to not be worth it.

**Comment by Manishearth** [comment]
I think being able to run multiple cargo builds at the same time should be worth it. That's a pretty major use case.

**Comment by eminence** [comment]
I think I'm running into a situation where having `cargo run` running in one window, prevents `cargo build` from working another window (it's blocked on this waiting for file-lock message).  Is that expected?

**Comment by alexcrichton** [comment]
@eminence yeah if there's a git dependency then this would cause those two commands to have to wait for one another.

**Comment by bozaro** [comment]
In my case I start building on Jenkins two copies of same project in differ working copy builder and get deadlock.

For example:

First build job (duration 1h 57m):
```
[octobuild (release)] Running shell script
+ cargo test
    Blocking waiting for file lock on the registry index
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Updating git repository `https://github.com/bozaro/nickel.rs.git`
    Updating git repository `https://github.com/bozaro/rust-crypto.git`
    Blocking waiting for file lock on the git checkout
Sending interrupt signal to process
Sending interrupt signal to process
    Updating git repository `https://github.com/bozaro/ipc-rs.git`
```

Second build job (duration 1h 57m):
```
[octobuild (release)@2] Running shell script
+ cargo build --release --target i686-pc-windows-gnu
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Blocking waiting for file lock on the registry index
    Updating git repository `https://github.com/bozaro/ipc-rs.git`
    Blocking waiting for file lock on the git database
    Blocking waiting for file lock on the git checkout
Sending interrupt signal to process
Sending interrupt signal to process
```

As workaround I simply disable parallel build execution.

**Comment by alexcrichton** [comment]
@bozaro what version of Cargo was that? In theory that's fixed on master, but there may be something lurking.

**Comment by bozaro** [comment]
Cargo version:
```
cargo 0.12.0-nightly (6b98d1f 2016-07-04)
```

**Comment by bozaro** [comment]
Related commit: cd8ad10fe93779abd608aa4185317473f874832f

**Comment by alexcrichton** [comment]
Ah ok, I think this may have been fixed with https://github.com/rust-lang/cargo/pull/2991, does nightly Rust work?

**Comment by bozaro** [comment]
I'm failer to build with nightly rust because strange rustup error (error on installing cross-compile target). But nobody promised that the night version is stable, so that is OK.

I migrated to today released rust 1.12.0 and looks like it don't have a cargo deadlock,  but compile projects successively one-by-one.

**Comment by Manishearth** [comment]
> But nobody promised that the night version is stable, so that is OK.

You should still file bugs for regressions. Though this is a rustup error, not a nightly error, so file it on the rustup repo.

**Comment by alexcrichton** [comment]
Ok, thanks @bozaro!

**Comment by cardoe** [comment]
@alexcrichton In the same thread that @eminence mentioned. If I run `cargo run` and it has completed the build and is executing the binary, if you use git checkouts in `Cargo.toml` you will get a block waiting on git checkouts. However the binary is running and there is no manipulation with the checkouts so that block seems invalid.

**Comment by alexcrichton** [comment]
@cardoe seems bad! Want to open a separate issue about that?

**Comment by matklad** [comment]
Hi, @SimonSapin can you check if the issue is present with the current Cargo? I think that it might be fixed in https://github.com/rust-lang/cargo/pull/3213.

Now Cargo checksout each revision to a separate directory, and holds the lock only during checkout.

**Comment by SimonSapin** [comment]
Sorry I forgot about this issue. Yes, this is fixed in recent Cargo. I’m back to using a single shared `$CARGO_HOME` for multiple checkouts of Servo, and they can compile at the same time.

Thanks @matklad for your work in #3213!
